### PR TITLE
[SYCL][CUDA] Add group algorithms

### PR DESCRIPTION
**Pull request by Pennycook**
Adds support for the following SPIR-V instructions to libclc:
- OpGroupAll, OpGroupAny
- OpGroupBroadcast
- OpGroupIAdd, OpGroupFAdd
- OpGroupFMin, OpGroupUMin, OpGroupSMin
- OpGroupFMax, OpGroupUMax, OpGroupSMax

At sub-group scope, these operations employ shuffles and other warp
instructions.

At work-group scope, partial results from each sub-group are combined
via shared memory.

The current implementation reserves 512 bytes of shared memory for any kernel
using a group algorithm, which is sufficient to cover the worst case.
Determining the correct amount of shared memory to reserve for a specific
kernel will likely require a dedicated compiler pass.

Signed-off-by: John Pennycook <john.pennycook@intel.com>

### CODE REVIEW

**Comment by Pennycook** [comment]
Apologies to reviewers for the giant merge request.  I've split things out into multiple commits to allow the libclc changes to be reviewed separately from the changes to the tests, but there wasn't an obvious way to further subdivide the libclc changes.

All of the SPIR-V functionality is defined in terms of one mega instruction: whether it computes a reduction, inclusive scan or exclusive scan is controlled by one parameter; and whether the operation is performed at sub-group or work-group scope is another parameter.  The DPC++ implementation of scan depends on broadcast, and so I ended up having to implement everything at once.

**Comment by jbrodman** [comment]
Very nice!

Are there any corner cases that don't work? Do we need to document what is/is not expected to work?

**Comment by Pennycook** [comment]
> Are there any corner cases that don't work?

Not intentionally.  It's passing all the tests, and I tried a few extra cases as a sanity check (e.g. non-power-of-2 sub-groups, weird data types) and things seemed to work.  I haven't done exhaustive testing, though, so there might still be some bugs.

> Do we need to document what is/is not expected to work?

Yeah.  I guess we should add "CUDA" to the appropriate rows in https://github.com/intel/llvm/tree/sycl/sycl/doc/extensions#extensions?

File: libclc/ptx-nvidiacl/libspirv/group/collectives.cl [hunk 1]
```diff
@@ -0,0 +1,419 @@
...
+__CLC_SUBGROUP_SHUFFLE_I32(uint);
+#undef __CLC_SUBGROUP_SHUFFLE_I32
+
+_CLC_DEF _CLC_OVERLOAD _CLC_CONVERGENT ulong __clc__SubgroupShuffle(ulong x,
+                                                                    uint idx) {
+  uint lo = (uint)x;
+  uint hi = (uint)(x >> 32);
+  lo = __nvvm_shfl_sync_idx_i32(__clc__membermask(), lo, idx, 0x1f);
+  hi = __nvvm_shfl_sync_idx_i32(__clc__membermask(), hi, idx, 0x1f);
+  return (((ulong)hi) << 32ul) | lo;
```
Comment (bader) [inline libclc/ptx-nvidiacl/libspirv/group/collectives.cl:67]: You might consider casting `ulong` to `uint2` to improve readability.
```suggestion
  uint2 y = as_uint2(x);
  y.lo = __nvvm_shfl_sync_idx_i32(__clc__membermask(), y.lo, idx, 0x1f);
  y.hi = __nvvm_shfl_sync_idx_i32(__clc__membermask(), y.hi, idx, 0x1f);
  return as_ulong(y);
```

The same applies to `__clc__SubgroupShuffleUp`.

File: libclc/ptx-nvidiacl/libspirv/group/collectives_helpers.ll [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+}
+
+define i8 addrspace(3)* @__clc__get_group_scratch_char() nounwind alwaysinline {
+entry:
+  %ptr = getelementptr inbounds [64 x i64], [64 x i64] addrspace(3)* @__clc__group_scratch, i64 0, i64 0
+  %cast = bitcast i64 addrspace(3)* %ptr to i8 addrspace(3)*
+  ret i8 addrspace(3)* %cast
+}
+
+define i8 addrspace(3)* @__clc__get_group_scratch_short() nounwind alwaysinline {
```
Comment (bader) [inline libclc/ptx-nvidiacl/libspirv/group/collectives_helpers.ll:21]: i8 -> i16?

**Review by bader — COMMENTED** [review]
> > Do we need to document what is/is not expected to work?
> 
> Yeah. I guess we should add "CUDA" to the appropriate rows in https://github.com/intel/llvm/tree/sycl/sycl/doc/extensions#extensions?

Let's update the documentation in this PR.

File: libclc/ptx-nvidiacl/libspirv/group/collectives.cl (hunk 1 above)
Comment (Pennycook) [reply libclc/ptx-nvidiacl/libspirv/group/collectives.cl]: Ooh, that's much nicer.  Thanks.  Fixed in 3b044955bb9994a3ea6e3dc5efbbee9a6bdc5f24.

File: libclc/ptx-nvidiacl/libspirv/group/collectives_helpers.ll (hunk 2 above)
Comment (Pennycook) [reply libclc/ptx-nvidiacl/libspirv/group/collectives_helpers.ll]: Fixed in 3375433c974f4446409e4585754f97abb8e6ab59.

File: sycl/test/group-algorithm/exclusive_scan.cpp [hunk 3]
```diff
@@ -13,8 +10,10 @@
 // unconditionally. Using operators specific for spirv 1.3 and higher with
 // -spirv-max-version=1.1 being set by default causes assert/check fails
 // in spirv translator.
-// RUNx: %clangxx -fsycl -fsycl-targets=%sycl_triple -DSPIRV_1_3 %s -o %t13.out
+// RUNx: %clangxx -fsycl -fsycl-targets=%sycl_triple -DSPIRV_1_3 %s -I . -o
+// %t13.out
```
[the remark this answers is not in the document]
Comment (Pennycook) [reply sycl/test/group-algorithm/exclusive_scan.cpp:14]: clang-format did this.  How should I override it?

[the remark this answers is not in the document]
Comment (bader) [reply sycl/test/group-algorithm/exclusive_scan.cpp:14]: There is a rule in our clang-format configuration file that should keep RUN commands intact, but according to my understanding this rule doesn't match "RUNx" command. See https://github.com/intel/llvm/blob/sycl/sycl/test/.clang-format#L2

Alternatives to applying my current suggestion (i.e. manually revert clang-format):
1. Remove lines 13-14. At this moment LIT do not run clang.
2. add `\` at the end of line 13 so, that when RUNx is replaced by RUN, the command won't be truncated.

@romanovvlad, FYI.

**Review by bader — DISMISSED** [review]
libclc part looks good to me. Thanks!
@Naghasan, FYI.

File: sycl/test/group-algorithm/exclusive_scan.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (Pennycook) [reply sycl/test/group-algorithm/exclusive_scan.cpp]: I like option 2. here.  It satisfies clang-format and we get to keep the lines around.  Fixed in a5a60193f7cecafa5e08556daf5158a451cbe229.

**Review by bader — APPROVED** [review]

**Review by v-klochkov — APPROVED** [review]

**Review by jbrodman — APPROVED** [review]
